### PR TITLE
Numpy bitwise_and operation added

### PR DESCRIPTION
**Pull request by zoeygxy**
## Description ##
Numpy element-wise operator [bitwise_and](https://docs.scipy.org/doc/numpy/reference/generated/numpy.bitwise_and.html#numpy.bitwise_and) added

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Added numpy element-wise operator: [bitwise_and](https://docs.scipy.org/doc/numpy/reference/generated/numpy.bitwise_and.html#numpy.bitwise_and) 

## Comments ##
@haojin2

### CODE REVIEW

File: src/operator/numpy/np_elemwise_binary_op.cc [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file np_elemwise_binary_op.cc
+ * \brief CPU Implementation for element-wise bitwise add.
```
Comment (haojin2) [inline src/operator/numpy/np_elemwise_binary_op.cc]: Fix comment.

File: src/operator/numpy/np_elemwise_binary_op.cc [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+ */
+
+
```
Comment (haojin2) [inline src/operator/numpy/np_elemwise_binary_op.cc]: nit: get rid of the extra blank line.

File: src/operator/numpy/np_elemwise_binary_op.cu [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+#include <mxnet/base.h>
+#include "../mshadow_op.h"
+#include "../tensor/elemwise_binary_broadcast_op.h"
+
+namespace mxnet {
+namespace op {
+
+NNVM_REGISTER_OP(_np_bitwise_and)
+.set_attr<FCompute>("FCompute<gpu>", BinaryBroadcastCompute<gpu, mshadow_op::bitwise_and>);
+
```
Comment (haojin2) [inline src/operator/numpy/np_elemwise_binary_op.cu]: nit: get rid of the extra blank line.

File: tests/python/unittest/test_numpy_op.py [hunk 4]
```diff
@@ -1017,7 +1017,58 @@ def hybrid_forward(self, F, x):
...
+        ((1, ),(1, )),
+        ((3, 0), (3, 0)),  # zero-dim shape
+        ((0, 1), (0, 1)),  # zero-dim shape
+        ((2, 0, 2), (2, 0, 2)),  # zero-dim shape
+        ((1, ), (3, )),  # broadcast
+        ((2, 3), (2, 1)),  # broadcast
+        ((1, 3), (2, 3)),  # broadcast
+        ((1, 3), (2, 0, 3)),  # broadcast to zero-dim shape
+        ((1, 0, 1), (3, 0, 1)), # broadcast of zero-dim shape
+        ((), ()),  # zero-size shape
```
Comment (haojin2) [inline tests/python/unittest/test_numpy_op.py]: actually `()` is zero-dim while any shape that contain a 0 is zero-size.

File: tests/python/unittest/test_numpy_op.py [hunk 5]
```diff
@@ -1017,7 +1017,58 @@ def hybrid_forward(self, F, x):
...
+
+            mx_out = np.bitwise_and(x1, x2)
+            np_out = _np.bitwise_and(x1.asnumpy(), x2.asnumpy())
+            assert_almost_equal(mx_out.asnumpy(), np_out, rtol=1e-3, atol=1e-5)
+
 
 if __name__ == '__main__':
     import nose
     nose.runmodule()
+
```
Comment (haojin2) [inline tests/python/unittest/test_numpy_op.py]: get rid of the 2 extra blank lines

File: tests/python/unittest/test_numpy_op.py [hunk 6]
```diff
@@ -1017,7 +1017,58 @@ def hybrid_forward(self, F, x):
             ret_mx = net(data_mx)
             assert same(ret_mx.asnumpy(), ret_np)
 
+@with_seed()
```
Comment (haojin2) [inline tests/python/unittest/test_numpy_op.py]: One more blank line above (2 blank lines between the test functions)

File: src/operator/numpy/np_elemwise_binary_op.cc [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+namespace mxnet {
+namespace op {
+
+NNVM_REGISTER_OP(_np_bitwise_and)
+.set_num_inputs(2)
+.set_num_outputs(1)
+.set_attr<mxnet::FInferShape>("FInferShape", BinaryBroadcastShape)
+.set_attr<nnvm::FInferType>("FInferType", ElemwiseType<2, 1>)
```
Comment (wkcn) [inline src/operator/numpy/np_elemwise_binary_op.cc]: I think we should check whether the input is interger.

Comment (haojin2) [reply src/operator/numpy/np_elemwise_binary_op.cc]: We actually plan to refactor all bitwise logic operators later when all of them are checked into numpy branch, and that check will be handled by the refactored code later. Right now this is just a workaround. This answer also applies to the other comment of yours in the other bitwise logic operator's PR.

File: tests/python/unittest/test_numpy_op.py [hunk 8]
```diff
@@ -1018,6 +1018,56 @@ def hybrid_forward(self, F, x):
...
+    for hybridize in [True, False]:
+        for shape in shapes:
+            x1_shape, x2_shape = shape
+
+            test_bitwise_and = TestBitwiseAnd()
+            if hybridize:
+                test_bitwise_and.hybridize()
+
+            x1 = rand_ndarray(x1_shape, dtype=_np.dtype(int)).as_np_ndarray()
+            x1.attach_grad()
```
Comment (haojin2) [inline tests/python/unittest/test_numpy_op.py]: get rid of `attach_grad` if you're not checking the grad at all.

File: src/operator/mshadow_op.h [hunk 9]
```diff
@@ -147,6 +147,8 @@ MXNET_BINARY_MATH_OP_NC(elu, a > DType(0) ? a :
 
 MXNET_BINARY_MATH_OP_NC(elu_grad, a > DType(0) ? DType(1) : DType(b + a));
 
+MXNET_BINARY_MATH_OP(bitwise_and, static_cast<int64_t>(a) & static_cast<int64_t>(b));
```
[the remark this answers is not in the document]
Comment (haojin2) [reply src/operator/mshadow_op.h]: After the later refactor we will not need this cast at all, right now this is just a workaround.

File: tests/python/unittest/test_numpy_op.py (hunk 8 above)
Comment (zoeygxy) [reply tests/python/unittest/test_numpy_op.py]: Fixed.

**Comment by zoeygxy** [comment]
Waiting for GCD PR's  INT_TYPE_SWITCH to merge. Will use it to restrict the input dtype.

**Comment by zoeygxy** [comment]
This operator will be directly moved to the master branch. See #16009.

**Pull request closed without merging**